lib/src/merged_tree.rs

```diff
@@ -970,6 +973,257 @@ impl Iterator for TreeDiffIterator<'_> {
+            }
+        }
+    }
+}
+
+impl Stream for TreeDiffStreamImpl<'_> {
+    type Item = (RepoPath, BackendResult<(MergedTreeValue, MergedTreeValue)>);
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut need_items = false;
+        while !(self.items.is_empty() && self.pending_trees.is_empty()) {
+            // Go through all pending tree futures and poll them.
+            let mut pending_index = 0;
+            while pending_index < self.pending_trees.len()
+                && (need_items
+                    || pending_index < self.max_concurrent_reads
+                    || self.items.len() < self.max_queued_items)
+            {
+                let (_, future) = &mut self.pending_trees[pending_index];
+                if let Poll::Ready(tree_diff) = future.as_mut().poll(cx) {
+                    let (dir, _) = self.pending_trees.remove(pending_index).unwrap();
+                    let key = DiffStreamKey::normal(dir);
+                    // Whenever we add an entry to `self.pending_trees`, we also add an Ok() entry
+                    // to `self.items`.
+                    let (before, after) = self.items.remove(&key).unwrap().unwrap();
+                    // If this was a transition from file to tree or vice versa, add back an item
+                    // for just the removal/addition of the file.
+                    if before.is_present() && !before.is_tree() {
+                        self.items
+                            .insert(key.clone(), Ok((before, Merge::absent())));
+                    } else if after.is_present() && !after.is_tree() {
+                        self.items.insert(
+                            DiffStreamKey::file_after_dir(key.path.clone()),
+                            Ok((Merge::absent(), after)),
+                        );
+                    }
+                    self.add_dir_diff_items(key.path, tree_diff);
+                } else {
+                    pending_index += 1;
+                }
+            }
+
+            // Now emit the first file, or the first tree that completed with an error
+            if let Some(entry) = self.items.first_entry() {
+                match entry.get() {
+                    Err(_) => {
+                        // File or tree with error
+                        let (key, result) = entry.remove_entry();
+                        return Poll::Ready(Some((key.path, result)));
+                    }
+                    Ok((before, after)) if !before.is_tree() && !after.is_tree() => {
+                        let (key, result) = entry.remove_entry();
+                        return Poll::Ready(Some((key.path, result)));
+                    }
+                    _ => {
+                        need_items = true;
+                        if !self.pending_trees.is_empty() {
+                            return Poll::Pending;
+                        }
```

Comment:
Perhaps, this can be `while let Some(entry) = ...` loop. `need_items` is effectively set only if the `pending_trees` is empty, and the `items` won't grow in that case.

Reply:
Oh, good point! The `needs_items` was needed in an old version of this code (it's been through a lot of revisions). I've delete the variable and changed it to a loop now.

Reply:
Maybe the outermost loop can be removed?

Reply:
Oops, I was too quick to merge this. I'll send a follow-up.